### ISSUE TITLE
Dart mojom.dart generator should allow you generate top-package-level mojom.dart files

### ISSUE THREAD

**Issue by apwilson** [domokit/mojo#731]
foo.mojom:
```
module baz;
import bar.mojom;
```

bar.mojom
```
module baz;
```

Currently foo.mojom.dart imports bar.mojom.dart as:
```
import package:mojom/baz/bar.mojom.dart;
```

This is bad because I don't have a mojom package and I don't know why it even chose that (apparently it's due to legacy reasons).

If you instead specify a package like this:

foo.mojom:
```
[DartPackage "fizz"]
module baz;
import bar.mojom;
```

bar.mojom
```
[DartPackage "fizz"]
module baz;
```

foo.mojom.dart imports bar.mojom.dart as:
```
import package:fizz/baz/bar.mojom.dart;
```

I want to be able to do something such that foo.mojom.dart imports bar.mojom.dart as:
```
import package:baz/bar.mojom.dart;
```

You *can* do that currently if you do this:

foo.mojom:
```
[DartPackage ""]
module baz;
import bar.mojom;
```

bar.mojom
```
[DartPackage ""]
module baz;
```

But I'm told that's a bug that will be soon fixed.

**Comment by johnmccutchan** [comment]
@zanderso

**Comment by zanderso** [comment]
tl;dr: Given the current set of high-level design choices, this is not feasible.

Currently, the public .mojom.dart files consumed by Flutter, etc. are vended by Dart's package manager "pub". The [DartPackage] annotations on .mojom module names encode the mapping from .mojom files to pub packages. We can't use the .mojom module name as the package name for two reasons. First, the .mojom files having the same module names don't map logically onto collections of code that it makes sense to distribute as a single pub package. Second, module names and pub package names are two different namespaces. I.e. it will happen that the mojom module name you want to use is already taken by a pub package and visa versa. I think there are already a few examples of this in the Mojo tree. We could potentially encode the mapping from .mojom files to pub packages somewhere else, but as long as .mojom.dart files are consumed from pub packages, we have to have the mapping somewhere.

As a hack, for in-tree tests, if a .mojom file doesn't have a [DartPackage] annotation, we assign it to the dummy pub package named 'mojom', which we've squatted on to avoid name conflicts. Since I think we're not making use of this anymore, we should probably just emit an error instead.

**Comment by apwilson** [comment]
I'm not sure I see why a generated mojom.dart file can't be at the base level of a package.

If I created a dart package called baz and I shared it around I don't see why clients shouldn't be able to import 'package:baz/foo.mojom.dart'  There's no conflict assuming I'm the one who 'owns' the baz project.

Why would we want to require a mojom import to be  'package:baz/somethingelse/foo.mojom.dart' ?

**Comment by apwilson** [comment]
Note that the title of this issue may be misleading and may not be the solution to what I really want.  I just want to be able to specify top-package-level mojom.dart files.

**Comment by zanderso** [comment]
There either are currently, or were in the past, two .mojom files with the same name in different .mojom modules that needed to go into the same Dart package, probably various `types.mojom` files in the mojo_services package. Arguably those files should go into different Dart packages, but the bindings generation logic has to accommodate the situation where they're not. We could make less of an effort to avoid file name conflicts, but then the next slew of issues filed against bindings generation will be about that instead. I think we're at a not too terrible place at the moment, so please tolerate this relatively minor inconvenience.
